### PR TITLE
undo_swaps 

### PR DESCRIPTION
**Pull request by KetpuntoG**
**Context:** The `Permute` operation uses many SWAP gates to be implemented. In [(#1324)](https://github.com/PennyLaneAI/pennylane/issues/1324) a study was proposed to make this operator native to `default_qubit` in order to make the circuit more efficient. Instead, I have created a transformation that removes all the SWAP gates from the circuit by playing with the wires of the operations. I think it is a more efficient and flexible solution. 

**Description of the Change:** 
With a single reading of the circuit from right to left, all SWAPS in the circuit can be eliminated.

**Example**:

Without `undo_swaps`:

```python
import pennylane as qml
dev = qml.device('default.qubit', wires=3)

@qml.qnode(dev)
def qfunc():
    qml.Hadamard(wires=0)
    qml.PauliX(wires=1)
    qml.SWAP(wires=[0,1])
    qml.SWAP(wires=[0,2])
    qml.PauliY(wires=0)
    return qml.expval(qml.PauliZ(0))

print(qml.draw(qfunc)())
```
 0: ──H──╭SWAP──╭SWAP──Y──┤ ⟨Z⟩ 
 1: ──X──╰SWAP──│────────┤     
 2: ──────────╰SWAP─────┤  

With `undo_swaps`:

```python
@qml.qnode(dev)
@qml.transforms.undo_swaps
def qfunc():
    qml.Hadamard(wires=0)
    qml.PauliX(wires=1)
    qml.SWAP(wires=[0,1])
    qml.SWAP(wires=[0,2])
    qml.PauliY(wires=0)
    return qml.expval(qml.PauliZ(0))

print(qml.draw(qfunc)())
```

 0: ──Y──┤ ⟨Z⟩ 
 1: ──H──┤     
 2: ──X──┤

**Benefits:** 
The computational cost of the circuit calculation is reduced.

**Related GitHub Issues:**
[(#1324)](https://github.com/PennyLaneAI/pennylane/issues/1324)

### CODE REVIEW

**Comment by josh146** [comment]
very nice solution!

**Comment by KetpuntoG** [comment]
> very nice solution!

Thanks @josh146 😊

**Comment by KetpuntoG** [comment]
[sc-11831]

File: pennylane/transforms/optimization/undo_swaps.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                    )
+
+            else:
+                swap_wires_0, swap_wires_1 = current_gate.wires
+                map_wires[swap_wires_0], map_wires[swap_wires_1] = (
+                    map_wires[swap_wires_1],
+                    map_wires[swap_wires_0],
+                )
+            list_copy.pop(0)
+            continue
```
Comment (dime10) [inline pennylane/transforms/optimization/undo_swaps.py:100]: I think you can remove the continue here as it's the last statement in the loop.

File: tests/transforms/test_optimization/test_undo_swaps.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+        transformed_qfunc = undo_swaps(qfunc)
+
+        tape = qml.transforms.make_tape(transformed_qfunc)()
+        res = tape.execute(qml.device("default.qubit", wires=2))
+        assert len(tape.operations) == 2
+        assert np.allclose(res[0][0], 0.5)
+
+    def test_two_qubits_gates_transform(self):
+        """Test that a two-qubit gate change correctly with a SWAP."""
```
Comment (dime10) [inline tests/transforms/test_optimization/test_undo_swaps.py:44]:
```suggestion
        """Test that a two-qubit gate changes correctly with a SWAP."""
```

**Review by dime10 — CHANGES_REQUESTED** [review]
Hey @KetpuntoG, thank you for contributing this transform, great idea for an optimization :)

A few minor comments below.

File: tests/transforms/test_optimization/test_undo_swaps.py [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+
+        def qfunc():
+            qml.PauliX(wires=1)
+            qml.CNOT(wires=[0, 1])
```
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply tests/transforms/test_optimization/test_undo_swaps.py:48]: the thing is that the PauliX is going to change to one (assuming it passes the above test) 😄

[the remark this answers is not in the document]
Comment (dime10) [reply tests/transforms/test_optimization/test_undo_swaps.py:48]: But only on the target wire right, not the control wire?

File: pennylane/transforms/__init__.py [hunk 4]
```diff
@@ -131,6 +131,7 @@
     single_qubit_fusion,
     merge_amplitude_embedding,
     remove_barrier,
+    undo_swaps,
```
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply pennylane/transforms/__init__.py:137]: Oh that's right! Josh is going to kill me haha

File: pennylane/transforms/optimization/undo_swaps.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+# pylint: disable=too-many-branches
+from pennylane import apply
+from pennylane.transforms import qfunc_transform
+from pennylane.wires import Wires
+from pennylane.tape import stop_recording
+
+
+@qfunc_transform
+def undo_swaps(tape):
+    """Quantum function transform to remove SWAP gates. To do this it runs from right to left through the circuit changing the position of the qubits accordingly.
```
Comment (dime10) [inline pennylane/transforms/optimization/undo_swaps.py:25]:
```suggestion
    """Quantum function transform to remove SWAP gates. To do this it runs from right
    to left through the circuit changing the position of the qubits accordingly.
```
Probably want to wrap this a bit :)

File: tests/transforms/test_optimization/test_undo_swaps.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (KetpuntoG) [reply tests/transforms/test_optimization/test_undo_swaps.py:48]: I mean, we have:
-----C----SWAP
X----X-----SWAP

what I want to check is that it does not transform itself to:

  X---C---
    ---X----

[the remark this answers is not in the document]
Comment (dime10) [reply tests/transforms/test_optimization/test_undo_swaps.py:48]: Ah ok, I understand now :D

**Review by dime10 — APPROVED** [review]

File: pennylane/transforms/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/transforms/__init__.py:137]: 👀

**Pull request merged**